File: vulcan-web/src/crypto/wallet.ts
import * as elliptic from 'elliptic';
import CryptoJS from 'crypto-js';
import { Transaction, TransactionPayload } from '../types';

const ec = new elliptic.ec('secp256k1');

export const generateKeyPair = (): { privateKey: string; address: string } => {
  const keyPair = ec.genKeyPair();
  const privateKey = keyPair.getPrivate('hex');
  const publicKey = keyPair.getPublic('hex');
  
  return {
    privateKey,
    address: publicKey
  };
};

export const signData = (data: string, privateKeyHex: string): string => {
  const keyPair = ec.keyFromPrivate(privateKeyHex, 'hex');
  const dataHash = CryptoJS.SHA256(data).toString();
  const signature = keyPair.sign(dataHash);
  
  return signature.toDER('hex');
};

export const createTransactionData = (tx: TransactionPayload, timestamp: string): string => {
  return `${tx.from}${tx.to}${tx.amount}${tx.fee}${timestamp}`;
};

export const signTransaction = (
  tx: TransactionPayload,
  privateKeyHex: string
): Transaction => {
  const timestamp = new Date().toISOString();
  
  const dataToSign = createTransactionData(tx, timestamp);

  const hash = CryptoJS.SHA256(dataToSign).toString();
  
  const signature = signData(hash, privateKeyHex);
  
  const signedTx: Transaction = {
    id: '',
    from: tx.from,
    to: tx.to,
    amount: tx.amount,
    fee: tx.fee,
    signature: signature,
    timestamp: timestamp
  };
  
  const txData = `${signedTx.from}${signedTx.to}${signedTx.amount}${signedTx.fee}${signedTx.signature}${signedTx.timestamp}`;
  signedTx.id = CryptoJS.SHA256(txData).toString();
  
  return signedTx;
};

export const verifyKeyPair = (privateKeyHex: string, address: string): boolean => {
  try {
    const keyPair = ec.keyFromPrivate(privateKeyHex, 'hex');
    const publicKey = keyPair.getPublic('hex');
    return publicKey === address;
  } catch (error) {
    return false;
  }
};

export const formatAddress = (address: string): string => {
  if (address.length <= 16) return address;
  return `${address.slice(0, 8)}...${address.slice(-8)}`;
};

export const isValidAddress = (address: string): boolean => {
  return /^[0-9a-fA-F]{130}$/.test(address);
};

export const isValidPrivateKey = (privateKey: string): boolean => {
  return /^[0-9a-fA-F]{64}$/.test(privateKey);
};

File: wallet/wallet.go
package wallet

import (
	"crypto/ecdsa"
	"fmt"

	"github.com/OhMyDitzzy/vulcan/types"
)

// Wallet represents a user's wallet with key pair and address.
// In our blockchain, a wallet manages the user's private key and provides
// methods for signing transactions and managing funds.
type Wallet struct {
	PrivateKey *ecdsa.PrivateKey
	PublicKey  *ecdsa.PublicKey
	Address    string
}

// NewWallet creates a new wallet with a freshly generated key pair.
// Generate a random private key and derive the public key and address.
func NewWallet() (*Wallet, error) {
	privKey, err := GenerateKeyPair()
	if err != nil {
		return nil, fmt.Errorf("failed to generate key pair: %w", err)
	}
	
	pubKey := &privKey.PublicKey
	address := PublicKeyToAddress(pubKey)
	
	return &Wallet{
		PrivateKey: privKey,
		PublicKey:  pubKey,
		Address:    address,
	}, nil
}

// FromPrivateKey creates a wallet from an existing private key hex string.
// Use this to restore wallets from backed-up private keys.
func FromPrivateKey(privKeyHex string) (*Wallet, error) {
	privKey, err := PrivateKeyFromHex(privKeyHex)
	if err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}
	
	pubKey := &privKey.PublicKey
	address := PublicKeyToAddress(pubKey)
	
	return &Wallet{
		PrivateKey: privKey,
		PublicKey:  pubKey,
		Address:    address,
	}, nil
}

// SignTransaction signs a transaction with the wallet's private key.
// Compute the transaction hash and sign it, then set the signature
// on the transaction object. This proves that the wallet owner authorized
// the transaction.
func (w *Wallet) SignTransaction(tx *types.Transaction) error {
	if tx.From != w.Address {
		return fmt.Errorf("transaction sender does not match wallet address")
	}
	
	// Get data to sign
	dataToSign := tx.DataToSign()
	
	// Sign the data
	signature, err := Sign(dataToSign, w.PrivateKey)
	if err != nil {
		return fmt.Errorf("failed to sign transaction: %w", err)
	}
	
	// Set signature on transaction
	tx.SetSignature(signature)
	
	return nil
}

// VerifyTransactionSignature verifies that a transaction signature is valid.
// Extract the public key from the sender address and verify the signature
// against the transaction data. This ensures the transaction hasn't been
// tampered with and was actually signed by the claimed sender.
func VerifyTransactionSignature(tx *types.Transaction) (bool, error) {
	if tx.IsCoinbase() {
		return true, nil
	}
	
	pubKey, err := AddressToPublicKey(tx.From)
	if err != nil {
		return false, fmt.Errorf("invalid sender address: %w", err)
	}
	
	// Get data that was signed
	dataToSign := tx.DataToSign()

	valid, err := Verify(dataToSign, tx.Signature, pubKey)
	if err != nil {
		return false, fmt.Errorf("signature verification failed: %w", err)
	}
	
	return valid, nil
}

// Export returns the wallet's private key and address for backup..
func (w *Wallet) Export() (privateKeyHex, address string) {
	return PrivateKeyToHex(w.PrivateKey), w.Address
}

// CreateAndSignTransaction is a convenience method that creates and signs a transaction.
// Build a new transaction with the provided parameters and sign it with
// the wallet's private key in one step.
func (w *Wallet) CreateAndSignTransaction(to string, amount, fee uint64) (*types.Transaction, error) {
	tx := types.NewTransaction(w.Address, to, amount, fee)
	if err := w.SignTransaction(tx); err != nil {
		return nil, fmt.Errorf("failed to sign transaction: %w", err)
	}
	return tx, nil
}